### PR TITLE
[SPARK-4449][Core]Specify port range in spark

### PR DESCRIPTION
**Pull request by nssalian**
Specify port range in spark
JIRA link: https://issues.apache.org/jira/browse/SPARK-4449

Goal: To add a port range to services
Design: (Based on the input and the suggestions in t #3314 and #5722):
1) Added variables maxPort and failedPorts to help the implementation
2) The maxPort was explicitly assigned to be startPort + maxRetries to avoid Retries being lesser or greater than the specified port range.
3) Added the failedPorts ArrayBuffer to catch the failedPorts (during retry) which are in the range of the maxPort - startPort ( see Random logic)
4) This failedPorts list will be checked and the tryPort will not attempt those ports again in the random..
5) If the randomized port does not belong to the failedPorts list and a privileged port, it will be tried. 
6) There’s a if block to check if there are sufficient ports left to attempt within the range (not sure if this is needed)

Requesting review.

### CODE REVIEW

**Comment by nssalian** [comment]
Jenkins, slow test please

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 1]
```diff
@@ -193,7 +187,7 @@ private[spark] object Utils extends Logging {
   }
 
   /**
-   * Primitive often used when writing [[java.nio.ByteBuffer]] to [[java.io.DataOutput]]
+   * Primitive often used when writing java.nio.ByteBuffer to java.io.DataOutput
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: Are these unrelated changes?

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 2]
```diff
@@ -2013,22 +2009,40 @@ private[spark] object Utils extends Logging {
    */
   def startServiceOnPort[T](
       startPort: Int,
+      maxPort: Int,
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: I don't see how this addresses the issue, since it just gives a hook into this method, but doesn't address how users would specify ranges, or how to deal with the new semantics. ALso why would the caller pass in `failedPorts`?

**Comment by andrewor14** [comment]
@nssalian can you explain the difference between this and #5722? IIUC this introduces some kind of global port range right? I think our past discussions with @srowen @vanzin and @tgravescs already ruled that solution out.

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 3]
```diff
@@ -2017,18 +2011,23 @@ private[spark] object Utils extends Logging {
...
       // Do not increment port if startPort is 0, which is treated as a special port
       val tryPort = if (startPort == 0) {
         startPort
-      } else {
-        // If the new port wraps around, do not try a privilege port
-        ((startPort + offset - 1024) % (65536 - 1024)) + 1024
+      }
+      else {
+        // If the new port wraps around, do not try a privilege port or a failedPort
+        while(!failedPorts.contains((startPort + Math.random() * (maxPort - startPort + 1))) &&
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: This is an array buffer, so contains is linear time, which is pretty expensive. Also this code is very difficult to understand.

**Comment by srowen** [comment]
Yes, do you mind closing this PR? this is not quite getting at what needs to happen for SPARK-4449

**Comment by nssalian** [comment]
Closing the PR.
